vcd/provider.go

```diff
@@ -108,6 +112,48 @@ func Provider() terraform.ResourceProvider {
 	}
 }
 
+// This is a global MutexKV for all resources
+var vcdMutexKV = mutexkv.NewMutexKV()
+
+func lockVapp(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("name").(string))
+	log.Printf("[TRACE] Acquiring lock for vapp with key %s.", key)
+	vcdMutexKV.Lock(key)
+	log.Printf("[TRACE] Locked vapp with key %s.", key)
+}
+
+func unLockVapp(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("name").(string))
+	log.Printf("[TRACE] Unlocked vapp with key %s.", key)
+	vcdMutexKV.Unlock(key)
+}
+
+func lockParentVapp(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("vapp_name").(string))
+	log.Printf("[TRACE] Acquiring lock for parent vapp with key %s.", key)
+	vcdMutexKV.Lock(key)
+	log.Printf("[TRACE] Locked parent vapp with key %s.", key)
+}
+
+func unLockParentVapp(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("vapp_name").(string))
+	log.Printf("[TRACE] Unlocked parent vapp with key %s.", key)
+	vcdMutexKV.Unlock(key)
+}
+
+func lockParentEdgeGtw(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|edge:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("edge_gateway").(string))
+	log.Printf("[TRACE] Acquiring lock for edge gtw with key %s.", key)
+	vcdMutexKV.Lock(key)
+	log.Printf("[TRACE] Locked parent edge gtw with key %s.", key)
+}
+
+func unLockParentEdgeGtw(d *schema.ResourceData) {
+	key := fmt.Sprintf("org:%s|vdc:%s|edge:%s", d.Get("org").(string), d.Get("vdc").(string), d.Get("edge_gateway").(string))
+	log.Printf("[TRACE] Unlocked parent edge gtw with key %s.", key)
```

Comment:
Nit: please use the full word "gateway", so it's easier to find things in the log.

Reply:
logs removed